Feature extend filename components

nikola-rados opened the pull request:
This resolves #59 by adding an extra check for **downscaled gridded observations**.  A new property was added that checks a new case for `project_id` and `product`.  Furthermore, the `is_multi_year` check has been factored out to check both `gcm_derivative` and `other`.

rod-glover commented on nchelpers/__init__.py:
```diff
@@ -782,6 +782,11 @@ def is_gridded_obs(self):
+    def is_downscaled_gridded_obs(self):
+        """True iff the content of the file is downscaled gridded observations"""
+        return self.is_other and self.product == 'downscaled gridded observations'
+
```
:+1:

rod-glover commented on nchelpers/__init__.py:
```diff
@@ -1637,27 +1642,27 @@ def _cmor_type_filename_components(
+                mip_table=tres_to_mip_table and
+                          tres_to_mip_table.get(self.time_resolution, None)
+            )
+
```
Good. You've pulled the multi-year case into a case independent of other conditions. Is this what you are thinking here, i.e., that multi-year is independent of all other ways of categorizing the file?

rod-glover replied:
Oh, I see by your comments in the PR above that you did. Good.

rod-glover commented on nchelpers/__init__.py:
```diff
@@ -1700,15 +1705,17 @@ def _cmor_type_filename_components(
+                components.update(
+                    downscaling_method=self.method_id
+                )
+
```
Alternatively, pull the case of `self.is_downscaled_gridded_obs` up to the level of `self.is_other`, leaving `self.is_other` for last as it is a more general condition than `self.is_downscaled_gridded_obs`. 

Does that make sense? Generally speaking, we want to prefer less nesting of conditions.

nikola-rados replied:
Yes it does!

rod-glover commented on nchelpers/__init__.py:
```diff
@@ -1700,15 +1705,17 @@ def _cmor_type_filename_components(
         elif self.is_other:
 
             # CAUTION: Temporary solution to a bigger problem here
```
Let's add a link to https://pcic.uvic.ca/confluence/display/CSG/Indexing+Gridded+Observation+Datasets here so that we can skip the search phase next time this kind of problem raises its head.

nikola-rados commented:
Let me know if I got it right this time @rod-glover!

nikola-rados commented:
As it turns out my idea of **downscaled gridded observations** is not actually a thing that we have to worry about.  I had misunderstood what had been done to the data I was working with.  I've removed the `is_downscaled_gridded_obs` property and check from the code since we will just be checking for **gridded observations** (for which a check already exists).  The `is_multi_year` check is still valid as it sticks the frequency into the `unique_id` and `cmor_filename`, so that check has been left where it is.  I've added `tiny_gridded_mClimSD_obs.nc` which contains a new case (gridded and multi year), and tests were implemented.

I've modified the original issue #59 slightly to reflect our findings.  @rod-glover since the code is slightly different from what you were reviewing before please let me know if there are any changes you want done!

rod-glover commented on tests/test_CFDataset.py:
```diff
@@ -282,6 +282,30 @@ def test_filepath(cwd, raw_dataset, converter, expected):
+     ('gridded_mClimSD_obs', 'is_streamflow_model_dgcm_output', False),
+     ('gridded_mClimSD_obs', 'is_streamflow_model_iobs_output', False),
+     ('gridded_mClimSD_obs', 'is_climdex_output', False),
+     ('gridded_mClimSD_obs', 'is_climdex_ds_gcm_output', False),
```
Add test for `is_gridded_obs` here and in all other tests for example files. A bit tedious, but we do need to test that it works correctly.

rod-glover commented on tests/test_CFDataset.py:
```diff
@@ -282,6 +282,30 @@ def test_filepath(cwd, raw_dataset, converter, expected):
+     ('gridded_mClimSD_obs', 'cmor_filename',
+      'pr_mClimSD_anusplin_historical_19710201-20000531.nc'),
+     ('gridded_mClimSD_obs', 'unique_id',
+      'pr_mClimSD_anusplin_historical_19710201-20000531'),
```
Yay!

nikola-rados commented:
@rod-glover I think this is ready to go, am I ok to merge?

rod-glover commented:
Good work. Merge this baby!
